api/client/webclient/webclient.go

```diff
@@ -48,20 +50,61 @@ func newWebClient(insecure bool, pool *x509.CertPool) *http.Client {
 	}
 }
 
-// Find fetches discovery data by connecting to the given web proxy address.
-// It is designed to fetch proxy public addresses without any inefficiencies.
-func Find(ctx context.Context, proxyAddr string, insecure bool, pool *x509.CertPool) (*PingResponse, error) {
-	clt := newWebClient(insecure, pool)
-	defer clt.CloseIdleConnections()
+// doWithFallback attempts to execute an HTTP request using https, and then
+// fall back to plain HTTP under certain, very specific circumstances.
+//  * The caller must specifically allow it via the allowPlainHTTP parameter, and
+//  * The target host must resolve to the loopback address.
+// If these conditions are not met, then the plain-HTTP fallback is not allowed,
+// and a the HTTPS failure will be considered final.
+func doWithFallback(ctx context.Context, clt *http.Client, allowPlainHTTP bool, method string, proxyAddr string, path string) (*http.Response, error) {
+	// first try https and see how that goes
+	endpoint := fmt.Sprintf("https://%s%s", proxyAddr, path)
+
+	log.Printf("Attempting %s %s", method, endpoint)
+	req, err := http.NewRequestWithContext(ctx, method, endpoint, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
Pass in the request as a parameter? It makes the method more flexible and cuts down the params you need to pass in (you can get the path from req.URL).

Reply:
Done.